Port remaining WebCore/loader types to the new serialization format

gavin-apple opened the pull request:
#### 148e18aedb4d65cfb801be5fef3af8267cd4c0cd
<pre>
Port remaining WebCore/loader types to the new serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=251434">https://bugs.webkit.org/show_bug.cgi?id=251434</a>
rdar://104866054

Reviewed by Alex Christensen.

This change includes porting types:
    - CanvasActivityRecord
    - AttributionSecondsUntilSendData
    - CrossOriginOpenerPolicyValue
    - CrossOriginOpenerPolicy
    - CustomHeaderFields
    - HTTPHeaderField
    - NavigationRequester
    - PolicyContainer
    - SessionHistoryVisibility
    - SubstituteData

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/AttributionSecondsUntilSendData.h:
(WebCore::PCM::AttributionSecondsUntilSendData::minSecondsUntilSend):
(WebCore::PCM::AttributionSecondsUntilSendData::encode const): Deleted.
(WebCore::PCM::AttributionSecondsUntilSendData::decode): Deleted.
* Source/WebCore/loader/CanvasActivityRecord.h:
(WebCore::CanvasActivityRecord::encode const): Deleted.
(WebCore::CanvasActivityRecord::decode): Deleted.
* Source/WebCore/loader/CrossOriginOpenerPolicy.h:
(WebCore::CrossOriginOpenerPolicy::encode const): Deleted.
(WebCore::CrossOriginOpenerPolicy::decode): Deleted.
* Source/WebCore/loader/CustomHeaderFields.h:
(WebCore::CustomHeaderFields::encode const): Deleted.
(WebCore::CustomHeaderFields::decode): Deleted.
* Source/WebCore/loader/HTTPHeaderField.cpp:
(WebCore::HTTPHeaderField::HTTPHeaderField):
(WebCore::HTTPHeaderField::isValid):
(WebCore::HTTPHeaderField::create):
* Source/WebCore/loader/HTTPHeaderField.h:
(WebCore::HTTPHeaderField::encode const): Deleted.
(WebCore::HTTPHeaderField::decode): Deleted.
* Source/WebCore/loader/NavigationRequester.h:
(WebCore::NavigationRequester::encode const): Deleted.
(WebCore::NavigationRequester::decode): Deleted.
* Source/WebCore/loader/PolicyContainer.h:
(WebCore::PolicyContainer::isolatedCopy):
(WebCore::PolicyContainer::encode const): Deleted.
(WebCore::PolicyContainer::decode): Deleted.
* Source/WebCore/loader/SubstituteData.h:
(WebCore::SubstituteData::SubstituteData):
(WebCore::SubstituteData::isValid const):
(WebCore::SubstituteData::shouldRevealToSessionHistory const):
(WebCore::SubstituteData::content const):
(WebCore::SubstituteData::mimeType const):
(WebCore::SubstituteData::textEncoding const):
(WebCore::SubstituteData::failingURL const):
(WebCore::SubstituteData::response const):
(WebCore::SubstituteData::encode const): Deleted.
(WebCore::SubstituteData::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/259811@main">https://commits.webkit.org/259811@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2768bfca13dfafe642d3bc749978eb356800047b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105963 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15028 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38805 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115150 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175270 "Failed to checkout and rebase branch from PR 9383") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109866 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16449 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6209 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98182 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114897 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111716 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12519 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95491 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40065 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94384 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27135 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81733 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8285 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28487 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8772 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5055 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14398 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48030 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6790 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10319 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/loader/SubstituteData.h:
```diff
@@ -32,74 +32,40 @@
+    bool isValid() const { return m_content != nullptr; }
+    bool shouldRevealToSessionHistory() const { return m_shouldRevealToSessionHistory == SessionHistoryVisibility::Visible; }
+
+    const FragmentedSharedBuffer* content() const { return m_content.get(); }
```
If you changed this return type to RefPtr and shouldRevealToSessionHistory to SessionHistoryVisibility you could avoid "friend struct IPC::ArgumentCoder"

achristensen07 commented on Source/WebCore/loader/HTTPHeaderField.cpp:
```diff
@@ -30,16 +30,31 @@
+    StringView strippedName = StringView(name).stripLeadingAndTrailingMatchedCharacters(RFC7230::isWhitespace<UChar>);
+    StringView strippedValue = StringView(value).stripLeadingAndTrailingMatchedCharacters(RFC7230::isWhitespace<UChar>);
+    
+    ASSERT(RFC7230::isValidName(strippedName) && RFC7230::isValidValue(strippedValue));
```
I don't think this assertion is valid when decoding from an untrusted source, even after stripping whitespace.

achristensen07 commented on Source/WebCore/loader/HTTPHeaderField.cpp:
```diff
@@ -30,16 +30,31 @@
-    String name = strippedName.length() == unparsedName.length() ? WTFMove(unparsedName) : strippedName.toString();
-    String value = strippedValue.length() == unparsedValue.length() ? WTFMove(unparsedValue) : strippedValue.toString();
-    return {{ WTFMove(name), WTFMove(value) }};
+std::optional<HTTPHeaderField> HTTPHeaderField::create(String&& unparsedName, String&& unparsedValue)
```
I think we should use CreateUsing=create and make a new attribute to allow the create function to return a std::optional, then just leave HTTPHeaderField class alone.

gavin-apple committed:
Port remaining WebCore/loader types to the new serialization format
https://bugs.webkit.org/show_bug.cgi?id=251434
rdar://104866054

Reviewed by Alex Christensen.

This change includes porting types:
    - CanvasActivityRecord
    - AttributionSecondsUntilSendData
    - CrossOriginOpenerPolicyValue
    - CrossOriginOpenerPolicy
    - CustomHeaderFields
    - HTTPHeaderField
    - NavigationRequester
    - PolicyContainer
    - SessionHistoryVisibility
    - SubstituteData

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/AttributionSecondsUntilSendData.h:
(WebCore::PCM::AttributionSecondsUntilSendData::minSecondsUntilSend):
(WebCore::PCM::AttributionSecondsUntilSendData::encode const): Deleted.
(WebCore::PCM::AttributionSecondsUntilSendData::decode): Deleted.
* Source/WebCore/loader/CanvasActivityRecord.h:
(WebCore::CanvasActivityRecord::encode const): Deleted.
(WebCore::CanvasActivityRecord::decode): Deleted.
* Source/WebCore/loader/CrossOriginOpenerPolicy.h:
(WebCore::CrossOriginOpenerPolicy::encode const): Deleted.
(WebCore::CrossOriginOpenerPolicy::decode): Deleted.
* Source/WebCore/loader/CustomHeaderFields.h:
(WebCore::CustomHeaderFields::encode const): Deleted.
(WebCore::CustomHeaderFields::decode): Deleted.
* Source/WebCore/loader/HTTPHeaderField.cpp:
(WebCore::HTTPHeaderField::HTTPHeaderField):
(WebCore::HTTPHeaderField::isValid):
(WebCore::HTTPHeaderField::create):
* Source/WebCore/loader/HTTPHeaderField.h:
(WebCore::HTTPHeaderField::encode const): Deleted.
(WebCore::HTTPHeaderField::decode): Deleted.
* Source/WebCore/loader/NavigationRequester.h:
(WebCore::NavigationRequester::encode const): Deleted.
(WebCore::NavigationRequester::decode): Deleted.
* Source/WebCore/loader/PolicyContainer.h:
(WebCore::PolicyContainer::isolatedCopy):
(WebCore::PolicyContainer::encode const): Deleted.
(WebCore::PolicyContainer::decode): Deleted.
* Source/WebCore/loader/SubstituteData.h:
(WebCore::SubstituteData::SubstituteData):
(WebCore::SubstituteData::isValid const):
(WebCore::SubstituteData::shouldRevealToSessionHistory const):
(WebCore::SubstituteData::content const):
(WebCore::SubstituteData::mimeType const):
(WebCore::SubstituteData::textEncoding const):
(WebCore::SubstituteData::failingURL const):
(WebCore::SubstituteData::response const):
(WebCore::SubstituteData::encode const): Deleted.
(WebCore::SubstituteData::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/259811@main
